test-data/unit/fine-grained.test

```diff
@@ -3458,3 +3458,121 @@ a.py:1: error: "int" not callable
 [delete nonexistent_stub.pyi.2]
 [out]
 ==
+
+[case testDunderNewUpdatedMod]
+import a
+[file a.py]
+import b
+
+b.C(int())
+[file b.py]
+class C:
+    def __new__(cls, x: int) -> C:
+        pass
+[file b.py.2]
+class C:
+    def __new__(cls, x: str) -> C:
+        pass
+[out]
+==
+a.py:3: error: Argument 1 to "C" has incompatible type "int"; expected "str"
+
+[case testDunderNewUpdatedFunc]
+import a
+[file a.py]
+import b
+def func() -> None:
+    b.C(int())
+[file b.py]
+class C:
+    def __new__(cls, x: int) -> C:
+        pass
+[file b.py.2]
+class C:
+    def __new__(cls, x: str) -> C:
+        pass
+[out]
+==
+a.py:3: error: Argument 1 to "C" has incompatible type "int"; expected "str"
+
+[case testDunderNewUpdatedMethod]
+import a
+[file a.py]
+import b
+class A:
+    def func(self) -> None:
+        b.C(int())
+[file b.py]
+class C:
+    def __new__(cls, x: str) -> C:
+        pass
+[file b.py.2]
+class C:
+    def __new__(cls, x: int) -> C:
+        pass
+[out]
+a.py:4: error: Argument 1 to "C" has incompatible type "int"; expected "str"
+==
+
+[case testDunderNewUpdatedSubclass]
+import a
+[file a.py]
+import b
+
+b.D(int())
+[file b.py]
+from c import C
+class D(C): pass
+[file c.py]
+class C:
+    def __new__(cls, x: str) -> C:
+        pass
+[file c.py.2]
+class C:
+    def __new__(cls, x: int) -> C:
+        pass
+[out]
+a.py:3: error: Argument 1 to "D" has incompatible type "int"; expected "str"
+==
+
+[case testDunderNewUpdatedAlias]
+import a
+[file a.py]
+import b
+
+b.D(int())
+[file b.py]
+from c import C
+D = C
+[file c.py]
+class C:
+    def __new__(cls, x: int) -> C:
+        pass
+[file c.py.2]
+class C:
+    def __new__(cls, x: str) -> C:
+        pass
+[out]
+==
+a.py:3: error: Argument 1 to "C" has incompatible type "int"; expected "str"
+
+[case testDunderNewUpdatedCallable]
+import a
+[file a.py]
+from typing import Callable, Any
+import b
+
+def func(arg: Callable[[int], Any]) -> None:
+    pass
+func(b.C)
+[file b.py]
+class C:
+    def __new__(cls, x: int) -> C:
+        pass
+[file b.py.2]
+class C:
+    def __new__(cls, x: str) -> C:
+        pass
+[out]
+==
```

Comment:
Ideas for additional test cases:
* Add `__new__` when originally there was nothing.
* Replace `__init__` with `__new__`.